### PR TITLE
Added HB+NS notification UDA functions

### PR DESCRIPTION
**Pull request by gvarao**

### CODE REVIEW

File: juniper_official/Paragon/Northstar/notify_northstar.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+node_get_url = 'http://ns-web.northstar:3301/NorthStar/API/v2/tenant/1/topology/1/nodes?q=$.[?(@.hostName %26%26 (@.hostName == "{hostname}"))]'
+link_get_url = 'http://ns-web.northstar:3301/NorthStar/API/v2/tenant/1/topology/1/links/?q=$.[?((@.endA%26%26(@.endA.node.id=="{node_id}")%26%26(@.endA.interfaceName=="{if_name}"))||(@.endZ%26%26(@.endZ.node.id=="{node_id}")%26%26(@.endZ.interfaceName=="{if_name}")))]'
+link_mnt_url = 'http://ns-web.northstar:3301/NorthStar/API/v2/tenant/1/topology/1/maintenances?q=$.[?(@.elements%26%26@.elements.filter(x => x.topoObjectType == "link" %26%26 x.index=="{link_index}").length >0 )]'
+mnt_url = 'http://ns-web.northstar:3301/NorthStar/API/v2/tenant/1/topology/1/maintenances'
+node_mnt_url = 'http://ns-web.northstar:3301/NorthStar/API/v2/tenant/1/topology/1/maintenances?q=$.[?(@.elements%26%26@.elements.filter(x => x.topoObjectType == "node" %26%26 x.index=={node_index}).length >0 )]'
+headers = {'Content-Type': 'application/json'}
+
+#Queries the NorthStar Node ID using device hostname.     
+def get_node_id(hostname, **kwargs):
+    r = requests.get(node_get_url.format(hostname=hostname), headers=headers, verify=False)
```
Comment (sriram-rsankar) [inline juniper_official/Paragon/Northstar/notify_northstar.py:17]: Should we call requests.get inside try except block ?

File: juniper_official/Paragon/Northstar/notify_northstar.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    else:
+        raise Exception("Node doesn't exist")
+
+
+#Queries the NorthStar link ID using device hostname and IFD name and IFL number. 
+def get_link_index(hostname, ifd_name, ifl_number, **kwargs):
+    node_params = get_node_id(hostname)
+    if node_params:
+        node_id = node_params[0]
+        r = requests.get(link_get_url.format(node_id=node_id, if_name=ifd_name +"."+ ifl_number), headers=headers, verify=False)
```
Comment (sriram-rsankar) [inline juniper_official/Paragon/Northstar/notify_northstar.py:32]: Please catch the exception raised by request.get.

File: juniper_official/Paragon/Northstar/notify_northstar.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        res = r.json()
+        if (r.status_code == 200) and (len(res)>0):
+            mnt_index = res[0]['maintenanceIndex']
+            res[0]['status']='completed'
+            res = json.dumps(res[0])
+            c = requests.put(mnt_url+'/'+str(mnt_index), data=res, headers=headers, verify=False)
+            print ('PUT  response='+str(c))
+        
+            d = requests.delete(mnt_url+'/'+str(mnt_index), headers=headers, verify=False)
+            print ('DELETE  response='+str(d))
```
Comment (sriram-rsankar) [inline juniper_official/Paragon/Northstar/notify_northstar.py:177]: Is this print needed

File: juniper_official/Paragon/Northstar/notify_northstar.py (hunk 1 above)
Comment (gvarao) [reply juniper_official/Paragon/Northstar/notify_northstar.py]: Added try except block

File: juniper_official/Paragon/Northstar/notify_northstar.py (hunk 2 above)
Comment (gvarao) [reply juniper_official/Paragon/Northstar/notify_northstar.py]: Added try except block

File: juniper_official/Paragon/Northstar/notify_northstar.py (hunk 3 above)
Comment (gvarao) [reply juniper_official/Paragon/Northstar/notify_northstar.py]: Removed print statements

**Comment by gvarao** [comment]
Hi Sri Ram,

I have addressed all comments.

Regards,
Venkat(GV)

**Review by sriram-rsankar — APPROVED** [review]

**Pull request merged**